Improve semantic token modifiers

0dinD opened the pull request:
This PR improves the semantic token modifiers in the following ways:

#### New token modifiers
- `transient`
- `volatile`
- `default`
- `synchronized`
- `native`
- `strictfp`
- `generic`
- `typeArgument` (Example: `String` and `Integer` in `Map<String, Integer>`)
- `importDeclaration` (Every simple name inside of an import declaration)

#### See this screenshot for an example of how to use the new modifiers:
![new-modifiers](https://user-images.githubusercontent.com/52179873/92304559-630e7b80-ef7f-11ea-9881-136f282839b8.png)

#### Performance improvements
- A bitmask is now used for the token modifiers, even before encoding.
  - Seems like an improvement over streaming them into an array, only to convert them back into a bitmask during encoding.
  - Makes it easier to combine token modifiers in steps, instead of doing it all at once.
- The old implementation sometimes called `SimpleName#resolveBinding()` twice for the same node, this is now avoided.
- An `int[]` is now used instead of `List<Integer>` for the encoded tokens, since the length of the array is known.

I do not have access to any fancy profiling tools, but from my testing the new implementation seems to be significantly faster than the old one, even with my additions to the token modifiers.

## Note
The only thing this PR is missing right now is some new tests for the new modifiers, which I will write when I have time. In the meantime, please verify this PR and let me know if there's anything that needs to change.

One thing that I would like your input on is what to call the `importDeclaration` token modifier. Is this a good name? Some other ideas that I had were: `imported`, `importTarget` or `importDeclarationTarget`. There is no standard token modifier for this.

fbricon commented:
@0dinD  `importDeclaration` is fine by me. @Eskibear can you please review this? I'm definitely not the semantic highlighting expert

Eskibear commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/TokenModifier.java:
```diff
@@ -31,54 +31,26 @@
-		}
-	};
+	// Standard Java modifiers
+	PUBLIC("public"),
```
I'm wondering whether we have to add **all** "standard Java modifiers", because ultimately we'll align modifiers with [LSP3.16](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_semanticTokens) as much as possible, via LSP4J. 

And massive modifiers would be potentially anonying (we did receive such feedback). I'm thinking, do people actually want to know below information from highlighting?
```
transient, volatile, default, synchronized, native, strictfp
```

My suggestion is to only keep important modifiers you are actually consuming now, also to avoid future breaking changes (when aligning to LSP) as best as we can.

0dinD replied:
Well, I do see your point here. The main purpose of this PR was to add the `generic`, `typeArgument` and `importDeclaration` modifiers. But I realised that adding the rest of the standard Java modifiers was easy and computationally cheap, so I thought "why not?". 

I'm not sure I understand what you mean by "align modifiers with LSP3.16"? We still need to have some custom modifiers that are not part of LSP3.16, and I don't see why these would be any different. Regarding the "massive modifiers", I think it would be rare to actually see all of those modifiers used together in a token, since many of them are mutually exclusive, or at least rarely used together. 

To be clear, I also think some of the standard Java modifiers are rarely going to be used, like `strictfp` for example. But I just didn't see any particular reason not to add the information, since it's trivial to check them in the same way as `public`, `static`, `readonly` etc. If you want me to remove the new "standard Java modifiers", I am also fine with that.

Eskibear replied:
I'm not that strongly opposed to that (as you mentioned you don't see any slow-down), my point is just "why to add them if they are rarely going to be used". It's indeed trivial to check them, it may cost tiny but still some time to check...

I still prefer not to add them here. One day if we do consume those modifiers, it's also trivial to add them back.

0dinD replied:
Yeah, I do see your point. I'll remove those modifiers from this PR in my next iteration, which I'll probably commit later this week. But can we keep the `native` modifier? I don't see much use for the other modifiers (`transient`, `volatile`, `default`, `synchronized` and `strictfp`), but `native` is something that I might actually use myself, as it's quite common to see when using native libraries.

Eskibear reviewed (commented):
Overall it looks good to me. Performance improvement looks promising. Just wondering whether we have to add **all** above modifiers.

0dinD commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/semantictokens/SemanticTokensLegend.java:
```diff
@@ -13,22 +13,28 @@
 public class SemanticTokensLegend {
-	private final List<String> tokenTypes;
-	private final List<String> tokenModifiers;
+	public static final SemanticTokensLegend INSTANCE = new SemanticTokensLegend();
```
Oh, I see the original purpose of SemanticTokenManager now. But I just thought it was a bit unnecessary after the changes I made, since it was only used to get the SemanticTokensLegend instance. I can re-implement the lazy-loading in SemanticTokensLegend instead.

But now that I think about it, is there any reason to keep the legend in-memory at all in Java-land? Couldn't we just create a new instance when requested by the client? Because the server never uses the legend, and the client only needs to request it once and then keep it in it's own memory. Does the server have any reason to keep holding a static instance of the legend then?

Eskibear replied:
>  is there any reason to keep the legend in-memory at all in Java-land?

It's a good finding. Indeed, we don't have to have the Legend, as it simply holds all possible values of `TokenType` and `TokenModifier`.

> Does the server have any reason to keep holding a static instance of the legend then?

When client talks to the server, `Legend` is their contract, and it was used to encode tokens. With consistency in mind, I just implemented in that way.

0dinD commented:
I have now addressed the review comments, and added some tests for the new modifiers. 

Additionally, I replaced the token assertion helper methods in `SemanticTokensCommandTest` with a new `TokenAssertionHelper` class using the builder pattern. It does the same thing, but is a lot nicer to work with since you don't have to figure out the line and column of every token. This is done by going through all provided tokens in order, and then asserting that the text at their position matches the expected text. Aside from making future tests easier to write, the assertion helper also provides more detailed error messages when the assertions fail.

@Eskibear what do you think of these changes? Have I missed anything in my new assertion helper? It's a big rewrite, but I think it's worth it in order to make the tests easier to write. I've already been able to add lots of new assertions with it that would have taken much more effort to write with the old assertion helper methods.

Eskibear reviewed (approved):
LGTM, thanks!

0dinD commented:
Great!  
Before this is merged, perhaps we should discuss whether or not to make #1514 a part of this PR? If we use the second alternative I presented I could add the `constructor` token to this PR.  
What do you think @fbricon @Eskibear?

Eskibear commented:
> add the constructor token to this PR

A constructor is a special function, so for me it's more reasonable to use `function` as token type, and add a `constructor` modifier (if you cannot find a proper one in [LSP pre-defined modifers](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_semanticTokens)).

Actually the idea doesn't make much sense to me. I don't see any use case where we want to distinguish `constructor` with `function` in color. E.g. 
```java
class Foo {
  public Foo() {
  }
}

Foo obj = new Foo(); // <--- You already know it's a constructor because of the `new` keyword before it.
```

fbricon commented:
While testing #1545, I noticed Java 15's `sealed` keyword is not supported. Can this PR fix it, or do you prefer I opened a separate ticket? 
<img width="184" alt="Screen Shot 2020-09-16 at 12 24 31 PM" src="https://user-images.githubusercontent.com/148698/93325616-ddf55300-f817-11ea-97b2-14113fdf93cf.png">

Eskibear commented:
Semantic highlighting doesn't cover keywords. Similar with `record` in Java 14, I'm afraid it has to be fixed in https://github.com/atom/language-java

Eskibear commented:
just opened a ticket https://github.com/atom/language-java/issues/232

0dinD commented:
@Eskibear I meant to say "`constructor` **modifier**", but I do agree that constructors should be considered methods rather than types. My point was that other users might disagree, and prefer to style constructors based on their type instead. By making it a modifier on the types, both sides of the argument could get what they want, either styling it via the type or just like a function with `*.constructor`.

But since I don't want the feature myself, and since it could create some other problems, I think I'll scrap the idea for now. If we get a feature request for it in the future, I know how to implement it.

@fbricon I haven't looked at the Eclipse JDT update for Java 15 yet, buy I'm pretty sure it should be possible by adding a new token type for keywords. That's something I was going to look at in a future PR, this one is just for some new token modifiers. 

@Eskibear I'm quite sure that there is a standard token type for keywords, buy the problem is that the Eclipse AST doesn't seem to report the document position of most keywords. However, there is an AST node for modifier keywords, so I'm pretty sure the `sealed` keyword is possible to implement.

Eskibear commented:
> I think I'll scrap the idea for now. If we get a feature request for it in the future, I know how to implement it.

👍 

>  AST doesn't seem to report the document position of most keywords.

Yes, that's one of the reasons we didn't highlight keywords here. Ideally, "static" content like keywords can be coverred by TextMate-based syntax highlighting (because it's cheap), and semantic highlighting is responsible for cases not coverred by those regular expressions, mainly for identifiers (of variables, classes, methods..etc).

I'm ok that we provide more tokens (e.g. some keywords) if it's possible, as long as it does no harm to the performance.

0dinD commented:
As far as performance goes, I think it's going to be pretty cheap. The main bottleneck with semantic highlighting seems to lie in computing the AST, but all we have to do in order to highlight keywords is to visit some additional nodes. And I've already improved the performance via this PR, so we have some more leeway in terms of smaller improvements like this.

0dinD commented:
@fbricon is this PR ready to merge now or should I add the tokens for modifier keywords as well?

fbricon commented:
@0dinD is you can easily add support for the new Java 15 keywords, then please go ahead

0dinD commented:
Alright, I'll add the keyword token type (and fix the conflicts) on Monday then.

fbricon commented:
see https://github.com/atom/language-java/commit/d99f04087892a5ff9c73465a2a20019aab04abdd if it helps

0dinD commented:
Thanks, but the semantic highlighting feature only depends upon the AST from Eclipse JDT. I've already checked that the new keywords are included in the AST, so implementing this token type should be quite trivial.

fbricon commented:
I noticed there's also a new `non-sealed` keyword in [JEP 360](https://openjdk.java.net/jeps/360). See also https://github.com/atom/language-java/issues/236

0dinD commented:
@fbricon yes, I've noticed that as well. Both the `sealed` and the `non-sealed` keywords are easy to implement in semantic highlighting, with the `Modifier` AST node. `yield` and `permits` are going to be harder since they are not their own AST nodes, and as such, their exact location is not available to the AST visitor. I have some workarounds in mind for adding all of the different keywords, but I think it deserves its own PR since it's going to require some more careful consideration and testing. (Also, I don't have time to work on it right now.)

I've now fixed the conflicts and added the `modifierKeyword` token type. It will look bad right now as it's a custom token with no styling, but that can be fixed with a PR similar to redhat-developer/vscode-java#1540. I will submit that PR soon.

fbricon commented:
@Eskibear any chance you can test this today? I'm planning on a release tomorrow

Eskibear commented:
I'm wondering whether we should directly use `modifier` type (a standard type in [LSP 3.16](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_semanticTokens)) instead of creating a new `modifierKeyword` type. If so, we don't bother to [declear the token](https://github.com/redhat-developer/vscode-java/pull/1636) in downstream Java extension. 

FYI, Related discussions: 
`modifier`: https://github.com/microsoft/language-server-protocol/issues/968#issuecomment-621026509
`modifierKeyword`: https://github.com/microsoft/vscode/issues/97063

0dinD commented:
@Eskibear oh, I didn't realise it was part of LSP, and only bothered to look at the VS Code docs. I think we should use `modifier` then, but the scope mapping would be necessary until that modifier has been added to VS Code.

Eskibear commented:
> I think we should use `modifier` then, but the scope mapping would be necessary until that token has been added to VS Code.

Looks good. In LSP3.16, now there are `modifier` and `keyword` types. To me, the first is for "modifier keywords", and the second is for other keyword tokens. 

As for the type `modifierKeyword`, I think we can just keep an eye on the discussion.

0dinD commented:
@Eskibear @fbricon I updated this PR and redhat-developer/vscode-java#1636, `modifierKeyword` --> `modifier`.

Eskibear reviewed (approved):
LGTM

fbricon commented:
Thanks @0dinD!
